server/chat.ts

```diff
@@ -1094,27 +1099,41 @@ export class CommandContext extends MessageContext {
 						this.errorReply(`Images are not allowed in personal rooms.`);
 						return null;
 					}
-					if (!/width=([0-9]+|"[0-9]+")/i.test(tagContent) || !/height=([0-9]+|"[0-9]+")/i.test(tagContent)) {
+					if (!/width ?= ?(?:[0-9]+|"[0-9]+")/i.test(tagContent) || !/height ?= ?(?:[0-9]+|"[0-9]+")/i.test(tagContent)) {
 						// Width and height are required because most browsers insert the
 						// <img> element before width and height are known, and when the
 						// image is loaded, this changes the height of the chat area, which
 						// messes up autoscrolling.
 						this.errorReply(`All images must have a width and height attribute`);
+						this.errorReply(`This image doesn't: <${tagContent}>`);
 						return null;
 					}
-					const srcMatch = /src\s*=\s*"?([^ "]+)(\s*")?/i.exec(tagContent);
+					const srcMatch = / src ?= ?"?([^ "]+)(?: ?")?/i.exec(tagContent);
 					if (srcMatch) {
 						if (!this.canEmbedURI(srcMatch[1])) return null;
+					} else {
+						this.errorReply(`All images must have a src attribute with no spaces in the URL`);
+						this.errorReply(`This image doesn't: <${tagContent}>`);
+						return null;
 					}
 				}
 				if (tagName === 'button') {
-					if (
-						(this.room.isPersonal || this.room.isPrivate === true) &&
-						!this.user.can('lock') && /<button[^>]/.test(htmlContent.toLowerCase().replace(/\s*style\s*=\s*"?[^"]*"\s*>/g, '>'))
-					) {
-						this.errorReply(`You do not have permission to use scripted buttons (buttons with attributes other than style="...") in HTML.`);
-						this.errorReply(`If you just want to link to a room, you can do this: <a href="/roomid"><button>button contents</button></a>`);
-						return null;
+					if ((this.room.isPersonal || this.room.isPrivate === true) && !this.user.can('lock')) {
+						const buttonName = / name ?= ?"([^"]*)"/i.exec(tagContent)?.[1];
+						const buttonValue = / value ?= ?"([^"]*)"/i.exec(tagContent)?.[1];
+						let legalButton = !buttonName;
+						if (buttonName === 'send' && buttonValue?.startsWith('/msg ')) {
+							const [pmTarget] = buttonValue.slice(5).split(',');
+							if (this.room.auth?.[toID(pmTarget)] === '*') {
+								legalButton = true;
+							}
+						}
+						if (!legalButton) {
+							this.errorReply(`You do not have permission to use most buttons. Here are the two types you're allowed can use:`);
+							this.errorReply(`1. Linking to a room: <a href="/roomid"><button>go to a place</button></a>`);
+							this.errorReply(`2. Sending a message to a Bot: <button name="send" value="/msg, BOT_USERNAME, MESSAGE">send the thing</button>`);
+							return null;
+						}
```

Comment:
I never liked the `legalButton` variable, and better error messages _would_ preclude it. Therefore:

```suggestion
						if (buttonName === 'send' && buttonValue?.startsWith('/msg ')) {
							const [pmTarget] = buttonValue.slice(5).split(',');
							if (this.room.auth?.[toID(pmTarget)] !== '*') {
								this.errorReply(`This button is not allowed: <${tagContent}>`);
								this.errorReply(`Your scripted button can't send PMs to ${pmTarget}, because that user is not a roombot.`);
								return null;
							}
						} else if (buttonName) {
							this.errorReply(`This button is not allowed: <${tagContent}>`);
							this.errorReply(`You do not have permission to use most buttons. Here are the two types you're allowed can use:`);
							this.errorReply(`1. Linking to a room: <a href="/roomid"><button>go to a place</button></a>`);
							this.errorReply(`2. Sending a message to a Bot: <button name="send" value="/msg, BOT_USERNAME, MESSAGE">send the thing</button>`);
							return null;
						}
```